### PR TITLE
Add `get_lock` api to `LockableCurrency`

### PR DESCRIPTION
**Pull request by shawntabrizi**
Currently, there is no way to inspect any existing locks using the `LockableCurrency` trait.

This adds such an API, and implements it in the balances and generic assets pallet.

Unfortunately, lock reasons in these pallets are lossy. From `WithdrawReasons` we convert to a smaller subset `Reasons`. Then when getting a lock, we can only do best effort conversion back from `Reasons` to `WithdrawReasons`.

Its not perfect, but certainly better than what we have today.

### CODE REVIEW

**Comment by xlc** [comment]
Fixes #4655?

**Comment by shawntabrizi** [comment]
@xlc looks like it. We could also add more apis if needed, for example to return all locks.

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -1247,6 +1262,21 @@ where
 {
 	type Moment = T::BlockNumber;
 
+	// Gets a lock by `id` on an account, returning `None` if there is no lock.
+	fn get_lock(
+		id: LockIdentifier,
+		who: &T::AccountId,
+	) -> Option<(T::Balance, WithdrawReasons)> {
+		Self::locks(who).into_iter()
+		.find_map(|l| {
```
Comment (gavofyork) [inline frame/balances/src/lib.rs:1271]: missing indent

File: frame/support/src/traits.rs [hunk 2]
```diff
@@ -867,6 +867,12 @@ pub trait LockableCurrency<AccountId>: Currency<AccountId> {
 	/// The quantity used to denote time; usually just a `BlockNumber`.
 	type Moment;
 
+	/// Retrieve lock information for `id`, returning `None` if no lock exists.
+	fn get_lock(
```
Comment (gavofyork) [inline frame/support/src/traits.rs:871]: this should probably take `(who: &AccountId, reasons: WithdrawReasons, maybe_exclude: Option<LockIdentifier>) -> Option<Self::Balance>`

**Comment by gavofyork** [comment]
> for example to return all locks

See my comment in the issue.

**Comment by shawntabrizi** [comment]
So if a pallet wants to know if it has placed a lock on a user, it needs to query twice (once with and once without the maybe_exclude filter)? Or keep track on its own?

That is essentially the question I am trying to answer with this API. "Does a lock with LockIdentifier exist on this account".

**Comment by ltfschoen** [comment]
I would appreciate it if there was the possibility of reopening this PR, as I'd like to use such an in-built `get_lock` function to 'read' storage values. I was going to use it precisely for the purpose that @shawntabrizi described: _"Does a lock with LockIdentifier exist on this account"_, as without the ability to 'read' storage values the API appears to only support CUD instead of CRUD.

File: frame/support/src/traits.rs (hunk 2 above)
Comment (ltfschoen) [reply frame/support/src/traits.rs:1087]: perhaps `reasons: Option<WithdrawReasons>`?

**Comment by kianenigma** [comment]
FWIW I think the current API of this PR is okay. Allowing for filtering based on a `reason` is nice, but end of the day can be done on the call site just as easily.

**Pull request closed without merging**